File: count-odd-numbers-in-an-interval-range/Solution1523.py
# https://leetcode.com/problems/count-odd-numbers-in-an-interval-range/

class Solution1523:

    # T: O(1), S: O(1)
    def countOdds(self, low: int, high: int) -> int:
        if (low & 1) == 0:
            low += 1
        return 0 if low > high else (high - low) // 2 + 1

    # T: O(h-l), S: O(1)
    def countOddsBrute(self, low: int, high: int) -> int:
        count = 0
        while low <= high:
            if (low & 1) == 1:
                count += 1
            low += 1
        return count


def main() -> None:
    sol = Solution1523()
    print(sol.countOdds(3, 7))  # 3
    print(sol.countOddsBrute(3, 7))  # 3
    print(sol.countOdds(8, 10))  # 1
    print(sol.countOddsBrute(8, 10))  # 1


if __name__ == "__main__":
    main()

File: count-odd-numbers-in-an-interval-range/Solution1523.java
// https://leetcode.com/problems/count-odd-numbers-in-an-interval-range/

class Solution1523 {

    // T: O(1), S: O(1)
    public int countOdds(int low, int high) {
        if ((low & 1) == 0) {
            low++;
        }
        return low > high ? 0 : (high - low) / 2 + 1;
    }

    // T: O(h-l), S: O(1)
    public int countOddsBrute(int low, int high) {
        int count = 0;
        while (low <= high) {
            if ((low & 1) == 1) {
                count++;
            }
            low++;
        }
        return count;
    }

    public static void main(String[] args) {
        var sol = new Solution1523();
        System.out.println(sol.countOddsBrute(3, 7)); // 3
        System.out.println(sol.countOdds(3, 7)); // 3
        System.out.println(sol.countOddsBrute(8, 10)); // 1
        System.out.println(sol.countOdds(8, 10)); // 1
    }

}
